security/pkg/server/ca/node_auth.go

```diff
@@ -15,28 +15,114 @@
 package ca
 
 import (
+	"context"
 	"fmt"
+	"sync"
 
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 
+	"istio.io/istio/pkg/cluster"
 	"istio.io/istio/pkg/kube"
 	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/multicluster"
+	"istio.io/istio/pkg/kube/namespace"
 	"istio.io/istio/pkg/security"
 	"istio.io/istio/pkg/spiffe"
+	"istio.io/istio/pkg/util/sets"
+	"istio.io/istio/security/pkg/server/ca/authenticate/kubeauth"
 )
 
 // NodeAuthorizer is a component that implements a subset of Kubernetes Node Authorization
 // (https://kubernetes.io/docs/reference/access-authn-authz/node/) for Istio CA. Specifically, it
 // validates that a node proxy which requests certificates for workloads on its own node is requesting
 // valid identities which run on that node (rather than arbitrary ones).
-type NodeAuthorizer struct {
-	trustedNodeAccounts map[types.NamespacedName]struct{}
+type NodeAuthorizer interface {
+	authenticateImpersonation(ctx context.Context, caller security.KubernetesInfo, requestedIdentityString string) error
+}
+
+// MulticlusterNodeAuthorizor is an implementation of NodeAuthorizer for multi-cluster environmnets.
+// please refer to https://docs.google.com/document/d/10uf4EvUVif4xGeCYQydaKh9Yaz9wpysao7gyLewJY2Q.
+// It is responsible for maintaining an index of node authenticators, one per cluster.
+// node authorizations from one cluster will be forwarded to the node authorizer for the same cluster.
+type MulticlusterNodeAuthorizor struct {
+	remoteNodeAuthenticators map[cluster.ID]*ClusterNodeAuthorizer
+	m                        sync.Mutex
+	filter                   namespace.DiscoveryFilter
+	trustedNodeAccounts      sets.Set[types.NamespacedName]
+}
+
+func NewMulticlusterNodeAuthenticator(filter namespace.DiscoveryFilter, trustedNodeAccounts sets.Set[types.NamespacedName],
+	addClusterHandler func(multicluster.ClusterHandler),
+) *MulticlusterNodeAuthorizor {
+	m := &MulticlusterNodeAuthorizor{
+		remoteNodeAuthenticators: map[cluster.ID]*ClusterNodeAuthorizer{},
+		filter:                   filter,
+		trustedNodeAccounts:      trustedNodeAccounts,
+	}
+	if addClusterHandler != nil {
+		addClusterHandler(m)
+	}
+	return m
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterAdded(cluster *multicluster.Cluster, _ <-chan struct{}) {
```

Comment:
nit: mNa/m, https://go.dev/wiki/CodeReviewComments#receiver-names

Reply:
Thank you for the reminder. I have renamed it from mNa to m